### ISSUE TITLE
[java] UnusedNullCheckInEquals FP in 6.31.0

### ISSUE THREAD

**Issue by testation21** [pmd/pmd#3099]
**Affects PMD Version:** 6.31.0

**Rule:**  UnusedNullCheckInEquals
https://pmd.github.io/pmd-6.31.0/pmd_rules_java_errorprone.html#unusednullcheckinequals

**Description:** 
There are two cases:
1) When there are two instance 
which are checked null referencing, 
It is safe to call `equals()` on any instance.
2) Also, when using `Objects.equals`, it shows alarms.

**Code Sample demonstrating the issue:**
```
class TestUnusedNullCheckInEquals{
        public void testPMD(){
                //case 1.
                if(obj1 != null && obj2 != null && obj2.equals(obj1));
                //case 2.
                if(obj1 != null && obj2 != null && Objects.equals(obj1, obj2));
        }
}
```

**Expected outcome:**
Don't show any alarm.

**Running PMD through:** *[CLI]*

**Comment by testation21** [comment]
I want to ask another suspicious case.

The documentation about the rule is as follow:
````
After checking an object reference for null, 
you should invoke equals() on that object rather than 
passing it to another object’s equals() method.
````

I understood it as using null checked object as caller is safe.
However, when we using static field of object 
such as Boolean.TRUE, Integer.SIZE and so on,
it is sure that it will not throw null pointer exception.

What do you think about these cases?
Are they false positive, or cases that have to be checked by developers?

Code example:
````
class TestUnusedNullCheckInEquals{
        public void TestPMD(){
                Boolean b = true;       
                if(b != null && Boolean.TRUE.equals(b));
        }
}
````

**Comment by adangel** [comment]
**case 1**

It's maybe not clear from the rule description, but in this example:

```java
if(obj1 != null && obj2 != null && obj2.equals(obj1));
```

the rule should report, that you should call equals on `obj1` and not `obj2`, since you checked obj1 already to be not null.

This essentially means, that the check `obj2 != null` is unnecessary and the check should be written `obj1.equals(obj2)`. Doing another check additionally, does no harm. This rule is in "error prone", so any reported violation should result in an error at runtime. But this is clearly not the case here.

So, yes - this case should probably not be reported by that rule.

There might be an opportunity for another rule to report unnecessary null checks.

Looking at the examples, you can trick the rule to ignore this case by reordering the checks like this:

```java
if(obj2 != null && obj2.equals(obj1) && obj1 != null);
```

But the check `obj1 != null` is really not needed here. If obj1 is null, then obj2.equals(null) will be false anyway.

**case 2**

```java
if(obj1 != null && obj2 != null && Objects.equals(obj1, obj2));
```

That's a similar case. Since obj1 is checked to be not null, the rule expects, that it is used in the following equals call. But this line also clearly doesn't throw a NPE and is not error prone, so that rule should not report it.

In this case, both null checks might be unnecessary (at least one of them for sure).

The rule suggests to write this line like this:

```java
if(obj1 != null && obj1.equals(obj2));
```

which is equivalent. But the former variant is also correct although more verbose.

**case 3**

```java
Boolean b = true;       
if(b != null && Boolean.TRUE.equals(b));
```

It's similar - the null check here is not necessary (so maybe another rule should report here a violation). But it is not error prone to double check for null here.

So, yes, that's also a false positive.

**Comment by jsotuyod** [comment]
I'm closing this as the issues were either invalid, have been previously fixed or have been already documented on #4907
